src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)  # LHS is just documentation here
+        # Don't need to store every `c`, need last + one for the next iteration
+    end
+    y = first(last(hobbits))
```

Comment:
This stores lots of tuples where really we just want one element, but cary the other. If you just pull `c` out of the closure, then it's very slow, I presume because of the famous closure bug. (Same idea in other places).

Reply:
Yes it might be, I have not fully wrapped my head around transducers. 

Note that in this rule it isn't just `unzipmap`, since the different iterations talk to each other. It would need some kind of `unzipaccumulate`.